enh(metatomic):  `nnpot` style implementation

HaoZeke opened the pull request:
cc @PicoCentauri 

Nicer implementation matching upstream. Performance is now around 22ns a day on `cosmopc5`, up from 14ns. Uses `vesin`, but follows the `nnpot` design.

Some notes:
- The notified pairlists are local with domain decomposition, so mapping to the model atom ordering is painful
  - NNPot throws `NotImplementedError` too for this

So the implementation currently:
- Gathers positions to main
- Runs the model
- Broadcasts forces

For instructions on optimal usage / reproduction consider [details here](https://github.com/HaoZeke/pixi_envs/tree/gromacs_metatomic/orgs/metatensor/gromacs).

Luthaf committed:
Intitial scaffolding of metatensor interface

HaoZeke committed:
[MAINT] Fix merge cruft

HaoZeke committed:
chore(metatensor): orchestrate vesin design

HaoZeke committed:
chore(build): try to find metatensor

HaoZeke committed:
chore(metatomic): trace nnpot through the code

HaoZeke committed:
chore(metatomic): register as an energy output

HaoZeke committed:
chore(metatomic): rename for the new ecosystem

HaoZeke committed:
chore(metatomic): regex rename for metatomic

PicoCentauri reviewed (commented):
Very cool. I took a quick look on one file. I will continue later.

HaoZeke commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.cpp:
```diff
@@ -0,0 +1,411 @@
+        // Run the model
+        auto ivalue_output = model_.forward({ std::vector<metatomic_torch::System>{ system },
+                                              eval_options_,
+                                              /*check_consistency=*/false });
```
One thing about this is if we want to have an explicit parameter or defer to the compilation mode (e.g. DEBUG mode compilations will automatically turn it on)

PicoCentauri replied:
Good question. In lammps we have it as an option. I am fine to just turn it on if we build with debug options.

HaoZeke replied:
Went with a parameter for now.

HaoZeke commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.cpp:
```diff
@@ -0,0 +1,411 @@
+    }
+
+    auto requested_output      = torch::make_intrusive<metatomic_torch::ModelOutputHolder>();
+    requested_output->per_atom = false;
```
Maybe a documentation note instead of a warning?

PicoCentauri replied:
Yes sounds good.

PicoCentauri commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.cpp:
```diff
@@ -0,0 +1,411 @@
+        GMX_THROW(InvalidInputError("Unsupported device requested: " + targetDevice));
+    }
+    GMX_LOG(logger_.info) << "Using device: " << device_.str();
+    model_.to(device_);
```
Ahh good good!

PicoCentauri reviewed (commented):
I think we also need the `TopologyPreprocessor` to remove the LJ, charges etc as the NNpot and the QMMM module is doing it.

Otherwise it looks already almost functional!

HaoZeke committed:
chore(metatomic): add topology processing

as is done in nnpot

HaoZeke committed:
maint(metatomic): docstring update and formatting

HaoZeke committed:
maint(params): use the right variables

HaoZeke committed:
doc(metatomic): add a note wrt tpr

Co-authored-by: PicoCentauri <PicoCentauri@users.noreply.github.com>

HaoZeke committed:
chore(metatomic): rework cmake dependencies

HaoZeke committed:
chore(metatomic): try to use more git and fallback

HaoZeke committed:
chore(metatomic): kang from LAMMPS

Co-authored-by: luthaf <luthaf@users.noreply.github.com>

HaoZeke committed:
chore(openmp): fix for cuda compilers

https://github.com/NVIDIA/cuda-samples/issues/333

HaoZeke committed:
chore(metatomic): resolve more compilation errors

HaoZeke committed:
chore(metatomic): try to fix compilers

Luthaf commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.cpp:
```diff
@@ -0,0 +1,411 @@
+
+    // 3. Determine device
+    std::string targetDevice = params_.device_;
+    if (targetDevice == "auto")
```
https://github.com/metatensor/metatomic/issues/21

HaoZeke committed:
chore(metatomic): grab vesin through cmake

HaoZeke committed:
chore(metatomic): remove caffe2 mkl workaround

HaoZeke committed:
chore(metatomic): easier imports

HaoZeke committed:
chore(metatomic): Forgotten include

HaoZeke committed:
chore(metatomic): forgotten cmake

HaoZeke committed:
feat(mta): rework a ton for compilation

Co-authored-by: PicoCentauri <PicoCentauri@users.noreply.github.com>

HaoZeke committed:
feat(mta): update vesin to 0.4.1

HaoZeke committed:
feat(mta): compile without errors

HaoZeke committed:
feat(mta): compile AND link correctly

HaoZeke commented:
First compiling build... (links too which is nice)

HaoZeke committed:
chore(mtaopt): rework in the style of nnpot

Co-authored-by: PicoCentauri <PicoCentauri@users.noreply.github.com>

HaoZeke committed:
chore(mdmta): register everywhere

HaoZeke committed:
tst(mtaopt): add reference tests for options

HaoZeke committed:
bug(mtaforce): ensure compatible extensions

HaoZeke committed:
bug(mtaforce): fix vesin calling conventions

HaoZeke committed:
chore(mtaforce): use the model dtype

HaoZeke committed:
chore(mtaforce): cleanup debugging

HaoZeke committed:
chore(mtapot): update for mpi usage and efficiency

HaoZeke committed:
chore(mtaforce): defensively handle dtypes

... and cleanup

HaoZeke committed:
feat(mtapot): use more internal helpers

HaoZeke commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.cpp:
```diff
@@ -0,0 +1,655 @@
+    {
+        // resize/allocate on non-main ranks and broadcast raw bytes of vector
+        nblock_abc(mpiComm_.isMainRank(), mpiComm_.comm(), static_cast<std::size_t>(n_atoms), &atomNumbers_);
+    }
```
This is a bit unclear to me, it works under the assumption that main knows all the atom numbers, something I can't seem to figure out.

HaoZeke committed:
chore(mtapot): cleanup a bit

HaoZeke committed:
chore(typo): fixup

Co-authored-by: PicoCentauri <PicoCentauri@users.noreply.github.com>

PicoCentauri commented:
This is the issue for reference.

https://gitlab.com/gromacs/gromacs/-/issues/5477#note_2876613421

Luthaf commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.h:
```diff
@@ -0,0 +1,115 @@
+    //! local copy of simulation box
+    matrix box_;
+    /// From EON
+    torch::jit::Module                                model_;
```
Please use `metatensor_torch::Module` instead, it will properly modify metatensor data in `to()`.

Luthaf commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.h:
```diff
@@ -0,0 +1,115 @@
+#include "gromacs/mdtypes/iforceprovider.h"
+
+#include "metatensor.hpp"
+#ifdef DIM
```
```suggestion
// both gromacs and torch define `DIM`, which result in a conflict. We don't need either, 
// so we undef before including headers.
#ifdef DIM
```

HaoZeke replied:
https://gitlab.com/gromacs/gromacs/-/issues/5490


and for good measure

https://github.com/pytorch/pytorch/issues/167718

Luthaf commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.cpp:
```diff
@@ -0,0 +1,564 @@
+
+    // Broadcast dtype info (so all ranks know whether the model uses float32 or float64).
+    // 0 = float64, 1 = float32
+    int dtype_code_local = 0;
```
torch::ScalarType is already an enum, can we not send that directly?

Luthaf replied:
@HaoZeke still relevant

Luthaf commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.cpp:
```diff
@@ -0,0 +1,564 @@
+{
+    // This function is called on domain decomposition.
+    // Its ONLY job is to update the `idxLookup_` map, which maps
+    // the contiguous ML atom index (0..n_atoms-1) to the
```
I don't understand what you mean here. If all ranks have access to all positions, then we can give all atoms to the model in each rank, and use `selected_atoms` to pass the sparse index/what the model in a given rank should be doing.

HaoZeke replied:
Needs some speed testing, this PR is way too slow as written.

Luthaf replied:
So, what is this function doing?

HaoZeke committed:
chore(cmake): rework for older versions and clean

Co-authored-by: Guillaume Fraux <luthaf@luthaf.fr>

HaoZeke committed:
chore(cpp): more cleanup

Co-authored-by: Guillaume Fraux <luthaf@luthaf.fr>

Update src/gromacs/applied_forces/metatomic/metatomic_options.cpp

Co-authored-by: Guillaume Fraux <luthaf@luthaf.fr>

Update src/gromacs/applied_forces/metatomic/metatomic_topologypreprocessor.h

Co-authored-by: Guillaume Fraux <luthaf@luthaf.fr>

Update src/gromacs/applied_forces/metatomic/metatomic_topologypreprocessor.cpp

Co-authored-by: Guillaume Fraux <luthaf@luthaf.fr>

PicoCentauri commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.cpp:
```diff
@@ -0,0 +1,564 @@
+            }
+            atomNumbers_[i] = options_.params_.atoms_.atom[gIdx].atomnumber;
+        }
+    }
```
We should also error here if an atomnumber is not known by the model.

PicoCentauri replied:
@Luthaf said it should be done but kinda I end up in an infinite loop when I run our argon test system with current PET-MAd which does not support argon...

PicoCentauri replied:
Okay, it is fine we error in metatomic

PicoCentauri commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.cpp:
```diff
@@ -0,0 +1,564 @@
+        auto torch_pbc = torch::tensor(std::vector<uint8_t>{ 1, 1, 1 },
+                                       torch::TensorOptions().dtype(torch::kBool).device(this->device_));
+        auto torch_types =
+                torch::tensor(atomNumbers_, torch::TensorOptions().dtype(torch::kInt32)).to(this->device_);
```
And I think then also the `atomNumbers_` will be different. I will check how we get them actually from the defined topology.

HaoZeke replied:
> Okay they come from the atomtypes directive in the [topolgy file](https://manual.gromacs.org/documentation/current/reference-manual/topologies/topology-file-formats.html#topology-file). I checked a couple of predefined topologies that live here:

Awesome.


> The only thing we might want to filter out with a warning is atom number 0 as they use as these are dummy particles for some forcefields

wouldn't it just error because they're Mwhatever? seems more reasonable? I'm not sure filtering them would be such a good idea.

PicoCentauri replied:
Yes maybe error is better. We do already I think in metatomic

HaoZeke committed:
Update src/gromacs/applied_forces/metatomic/metatomic_forceprovider.h

Co-authored-by: Guillaume Fraux <luthaf@luthaf.fr>

HaoZeke commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.cpp:
```diff
@@ -0,0 +1,564 @@
+        gromacs_scalar_type = torch::kFloat64;
+    }
+    // This blob_options is for reading GROMACS memory
+    auto blob_options = torch::TensorOptions().dtype(gromacs_scalar_type).device(torch::kCPU);
```
Data at this stage is always on CPU.

HaoZeke committed:
chore(mta): use - over _ consistently

HaoZeke committed:
chore(mta): minor cleanup

HaoZeke committed:
Merge branch 'main' into useVesin

HaoZeke committed:
chore(mta): handle pbcs the way nnpot does

HaoZeke commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider_stub.cpp:
```diff
@@ -0,0 +1,74 @@
+#include "gromacs/utility/basedefinitions.h"
+#include "gromacs/utility/exceptions.h"
+
+#include "metatomic_forceprovider.h"
```
This still needs a refactor.

HaoZeke commented:
OK I'm asking for another round of reviews on this, the two major problems are:
- Speed :: @PicoCentauri noted that this is far too slow, can't even get a step in at 3000 atoms..
- Build system :: The files need to be refactored a bit, at the moment, the stub isn't correct so it won't compile without `metatomic`

Why should this be merged then? @PicoCentauri and I looked at some small tests and it seems to be correct? Plus this way I don't have to worry about getting out of sync with main again.

PicoCentauri commented:
Thanks @HaoZeke for the effort!

Regarding speed: I will just run some more tests today on the current version. Maybe I find a studid bottleneck.

Regarding the build system: Probably @Luthaf is the master and he might help. I think we should probably do this before merging.

HaoZeke committed:
Merge remote-tracking branch 'glups/main' into useVesin

HaoZeke committed:
chore(mtatopo): rework topology as per nnpot

xref: https://gitlab.com/gromacs/gromacs/-/commit/7a30c21b289fbdd520166697ac0ae2a536fbd9cb

HaoZeke committed:
chore(bld): remove non-existent file

HaoZeke committed:
chore(bldi): try private linkage

HaoZeke committed:
chore(cmake): try a more explicit install

HaoZeke committed:
chore(mtatopo): fixup for more build issues

HaoZeke committed:
chore(mtapot): initialize and fix pointers

PicoCentauri commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.cpp:
```diff
@@ -0,0 +1,585 @@
+        }
+
+        // TODO(rg): determine device
+        model_.to(device_);
```
We can use the metatomic helper here that we wrote, no?

HaoZeke replied:
Yup

PicoCentauri commented on src/gromacs/applied_forces/metatomic/metatomic_options.cpp:
```diff
@@ -0,0 +1,279 @@
+ * \author Metatensor developers <https://github.com/metatensor>
+ * \ingroup module_applied_forces
+ */
+// TODO(rg): Figure out how to insert the model into the .tpr file
```
Yes I opened an issue about it. We just wait until they find a solution and use it :-)

Luthaf replied:
This will be a LOT harder to do once we have non-TorchScript models in metatomic

PicoCentauri commented on src/gromacs/applied_forces/metatomic/metatomic_options.cpp:
```diff
@@ -0,0 +1,305 @@
+    const int     numRegularAtoms = mtop->natoms - numMTAAtoms;
+
+    GMX_LOG(logger.info)
+            .appendText("Neural network potential interface is active, topology was modified!");
```
```suggestion
            .appendText("Metatomic interface is active, topology was modified!");
```

PicoCentauri commented on src/gromacs/applied_forces/metatomic/metatomic_options.cpp:
```diff
@@ -0,0 +1,305 @@
+    GMX_LOG(logger.info)
+            .appendText("Neural network potential interface is active, topology was modified!");
+    GMX_LOG(logger.info)
+            .appendTextFormatted("Number of embedded MTA atoms: %d\nNumber of regular atoms: %d\n",
```
```suggestion
            .appendTextFormatted("Number of embedded metatomic atoms: %d\nNumber of regular atoms: %d\n",
```

PicoCentauri commented on src/gromacs/applied_forces/metatomic/metatomic_options.h:
```diff
@@ -0,0 +1,136 @@
+class MpiComm;
+
+
+//! TODO
```
Do you want to fill this now?

PicoCentauri commented on src/gromacs/applied_forces/metatomic/metatomic_mdmodule.cpp:
```diff
@@ -0,0 +1,221 @@
+    std::unique_ptr<MetatomicForceProvider> force_provider_;
+};
+
+} // end anonymous namespace
```
```suggestion
} // end metatomic namespace
```

?

Luthaf replied:
no, this is an anonymous namespace

PicoCentauri commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider_stub.cpp:
```diff
@@ -0,0 +1,122 @@
+    ~MetatomicForceProvider();
+
+    /*! TODO
+     */
```
Should be maybe provide a small description.

PicoCentauri commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.h:
```diff
@@ -0,0 +1,126 @@
+#include "metatensor/torch.hpp"
+#include "metatomic/torch.hpp"
+#ifdef DIM
+// XXX(rg): ask gromacs folks to do GMX_DIM
```
Maybe link the [issue](https://gitlab.com/gromacs/gromacs/-/issues/5490) here

PicoCentauri commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.h:
```diff
@@ -0,0 +1,126 @@
+class MpiComm;
+
+/*! \brief \internal
+ * TODO
```
same as other todos

PicoCentauri commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.h:
```diff
@@ -0,0 +1,126 @@
+    MetatomicForceProvider(const MetatomicOptions&, const MDLogger&, const MpiComm&);
+    ~MetatomicForceProvider();
+
+    /*! TODO
```
same as other todo

PicoCentauri commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.h:
```diff
@@ -0,0 +1,126 @@
+                                                   bool                                 periodic);
+
+private:
+    /// From NNPot
```
maybe a better comment?

PicoCentauri commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.h:
```diff
@@ -0,0 +1,126 @@
+
+    //! local copy of simulation box
+    matrix box_;
+    /// From EON
```
```suggestion
```

PicoCentauri commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.cpp:
```diff
@@ -0,0 +1,564 @@
+        evaluations_options_->set_length_unit("nm");
+
+        auto outputs = capabilities_->outputs();
+        if (!outputs.contains("energy"))
```
You can put me in here. I can do it.

PicoCentauri reviewed (commented):
Code looks clean. I have some minor comments. Mostly about replacing some todos with actual code comments.

We should change the target branch from `main` to `metatomic` and set this do the default branch. This makes updating easier and we do the same as for lammps.

Luthaf commented on CMakeLists.txt:
```diff
@@ -64,7 +64,7 @@ endif()
 # The GROMACS convention is that these are the version number of the next
 # release that is going to be made from this branch.
-project(Gromacs VERSION 2026.0)
+project(Gromacs VERSION 2027.0)
```
?

HaoZeke replied:
? This is an upstream change..
https://github.com/metatensor/gromacs/commit/15e0cf6836042045d544a6d0bd273d965c6f7d84

Luthaf committed:
Make sure the code can compile both with an without metatomic

Luthaf commented on src/external/tng_io/src/tests/CMakeLists.txt:
```diff
@@ -50,7 +50,7 @@ if(TNG_BUILD_TEST)
 
 if(TNG_BUILD_EXAMPLES)
-    find_package(OpenMP)
+    find_package(OpenMP REQUIRED C CXX)
```
should we remove this change now that the branch is a bit more up to date?

Luthaf commented on src/gromacs/applied_forces/metatomic/metatomic_forceprovider.cpp:
```diff
@@ -0,0 +1,600 @@
+
+MetatomicForceProvider::~MetatomicForceProvider() = default;
+
+void MetatomicForceProvider::gatherAtomPositions(ArrayRef<const RVec> globalPositions)
```
So I don't understand this function. Do all rank have or have not access to all positions in gromacs? I though they did but this function looks like they don't.

HaoZeke replied:
The older papers suggests they did, but at some point this was changed. At the moment the ranks don't have all positions, and this is currently required AFAIK.

Luthaf replied:
Do we know what positions are available on each rank? Ideally we should not gather everything at every step if we can avoid it.

Also, how does gromacs tells us which atoms we should compute the energy for?

HaoZeke committed:
chore(cmake): don't test for mta stubs

HaoZeke committed:
bug(nnpotcmake): don't assume nvtx exists

HaoZeke committed:
chore(cmake): sync mta with nnpot enablers

HaoZeke committed:
feat(mtapot): add pick_device logic

HaoZeke committed:
chore(review): apply comments

Co-authored-by: PicoCentauri <PicoCentauri@users.noreply.github.com>

PicoCentauri reviewed (approved):
nice! I think this a very good first working version.

Luthaf commented on cmake/gmxManageNNPot.cmake:
```diff
@@ -66,8 +66,20 @@ if(NOT GMX_NNPOT STREQUAL "OFF")
+        endif()
+        
+        # TORCH_LIBRARIES contain imported target "torch" that will set all flags and include paths etc
+        list(APPEND GMX_COMMON_LIBRARIES ${_filtered_torch_libs})
```
not sure why there are changes to this file on this branch

HaoZeke replied:
To keep the build passing, I'll open this upstream too.

HaoZeke replied:
To keep the build passing, I'll open this upstream too.

PicoCentauri committed:
Added env variable for device picking

HaoZeke committed:
Merge pull request #2 from metatensor/device-env

Add env variable `GMX_METATOMIC_DEVICE` for device picking

HaoZeke committed:
feat(mtapot): 50% better performance..

HaoZeke committed:
chore(nnpot): cleaner fix

HaoZeke committed:
tst(mtaopt): fixup
